### ISSUE TITLE
Migrations and Nullable Reference Types

### ISSUE THREAD

**Issue by TorreyGarland** [aspnet/EntityFrameworkCore#18950]
I am generating migrations files for EF core with the new nullable reference types analysis turned on for an ASP.NET core app (3.0).

The migration file generated has several "new object[]" declarations that VS is flagging for CS8625 warnings (screenshot included).

![image](https://user-images.githubusercontent.com/1054548/69002447-96cbc800-08bd-11ea-8199-8aa6e1a2a368.png)

Is there a work-around or change coming so that we don't have to do a global suppression.  I want the nullable reference analysis, but just not for these migration files.

**Comment by ajcvickers** [comment]
@TorreyGarland Putting this in the backlog to consider in the future. The global suppression seems like the best workaround for now.

**Comment by Timovzl** [comment]
@ajcvickers @TorreyGarland Could either of you provide an example of such a global supression? I'm trying to get them to work, but with no success.

```
[assembly: SuppressMessage("Code Quality", "CS8625:Cannot convert null literal to non-nullable reference type.",
Justification = "EF migrations.",
Scope = "NamespaceAndDescendants", Target = "~N:My.Namespace.Migrations")]
```

I did read somewhere that the `NamespaceAndDescendants` scope was broken in VS 16.7, and that it would be fixed in 16.8. I'm now on 16.8.2, but the suppression is not helping.

**Comment by roji** [comment]
I think that by global suppression, it was meant that you don't activate the NRT feature at the project level (i.e. `<Nullable>disabled</Nullable>`). You can place your migrations in a separate project to keep nullability enabled in your actual code, if you want.

Another option is to place a `#nullable disable` at the top of each migration file.

**Comment by roji** [comment]
Clearing milestone to discuss if we want to do something for 6.0.

**Comment by Timovzl** [comment]
@roji I'm pretty sure global suppressions is generally the term for what I mentioned above, so I assume that's what @TorreyGarland meant. I can't see why it's not working, though.

Disabling nullable for the entire project is too impactful, unfortunately. Adding `#nullable disable` at the top of each migration file would work, as would disabling the warnings from those locations.

I'm still hoping for a solution that makes the problem go away, instead of requiring each developer to remember for each new migration. We have "treat warnings as errors" enabled in the DevOps pipeline, to prevent the solution from becoming riddled with warnings... so it's quite a nuisance. 😛

**Comment by roji** [comment]
You're right - global suppressions should indeed also work, I'm not sure why the above doesn't. I also agree we should provide a better solution than requiring developers to handle this themselves every time.

**Comment by maxkoshevoi** [comment]
How about adding `#nullable disable` at the top of each migration automatically as part of it's generation? 
Should be relatively easy to implement and would solve the problem completely without side effects.

**Comment by roji** [comment]
Yeah, that's what we did for scaffolding before 6.0. We should consider doing this (both for migrations and for the model snapshot).

**Comment by roji** [comment]
Opened #25624 to track adding `#nullable disable`, keeping this to track generating proper NRT code in the future.

**Comment by jacobjmarks** [comment]
I am currently facing a runtime exception regarding `null` usage in generated migration code for a code-first temporal table.
I have written the problem up on the below Issue and am referencing here as it may be relevant.
https://github.com/dotnet/EntityFramework.Docs/issues/3353#issuecomment-917859013

**Comment by roji** [comment]
@jacobjmarks you seem to be describing an actual (runtime) failure when running the Down migration, whereas this issue is about NRT annotation (which is a compile-time only thing). I suggest you open an issue  here (not in the docs repo) with precise instructions - if indeed we're scaffolding migration code which throws NullReferenceException, that's a bug.

/cc @maumar

**Comment by roji** [comment]
Scenarios:

* Seeding data (`new object[]`)
* Anonymous types when creating nullable Owned entities (from https://github.com/dotnet/efcore/issues/10695#issuecomment-950693625)
